_posts/2020-07-04-high-performance-scheduling.md

```diff
@@ -0,0 +1,344 @@
+costs.
+
+### Can we push this further?
+
+The first four stages of our pipeline happen on the client:
+
+1.  **Graph generation:** Some Python code in a Dask collection library, like
+    dask array, calls the sum function, which generates a task graph on the
+    client side.
+2.  **Graph Optimization:** We then optimize that graph, also on the client
+    side, in order to remove unnecessary work, fuse tasks, apply important high
+    level optimizations, and more.
+3.  **Graph Serializtion:** We now pack up that graph in a form that can be
+    sent over to the scheduler.
+4.  **Graph Communication:** We fire those bytes across a wire over to the
+    scheduler
+
+If we're able to stay with the high level graph representation through these
+stages all the way until graph communication, then we can communicate a far
+more compact representation up to the scheduler.  We can drop a lot of these
+costs, at least for the high level collection APIs (delayed and client.submit
+would still be slow, client.map might be ok though).
+
+This has a couple of other nice benefits:
+
+1.  User's code won't block, and we can alert the user immediately that we're
+    on the job
+2.  We've centralized costs in just the scheduler,
+    so there is now only one place where we might have to think about low-level code
+
+(some conversation here: https://github.com/dask/distributed/issues/3872)
+
+
+### However, low-level graph optimizations are going to be a problem
+
+In principle changing the distributed scheduler to accept a variety of graph
+layer types is a tedious but straightforward problem.  I'm not concerned.
+
+The bigger concern is what to do with low-level graph optimizations.
+Today we have three of these that really matter:
+
+1.  Task fusion: this is what keeps your `read_parquet` task merged with your
+    subsequent blockwise tasks
+2.  Culling: this is what makes `df.head()` or `x[0]` fast
+3.  Slice fusion: this is why `x[:100][5]` works well
+
+In order for us to transmit abstract graph layers up to the scheduler, we need
+to remove the need for these low level graph optimizations.  I think that we
+can do this with a combination of two approaches:
+
+
+#### More clever high level graph manipulation
+
+We already do this a bit with blockwise, which has its own fusion, and which
+removes much of the need for fusion generally.  But other blockwise-like
+operations, like `read_*` will probably have to join the Blockwise family.
+
+Getting culling to work properly may require us to teach each of the individual
+graph layers how to track dependencies in each layer type and cull themselves.
+This may get tricky.
+
+Slicing is doable, we just need someone to go in, grok all of the current
+slicing optimizations, and make high level graph layers for these
+computations.  This would be a great project for a sharp masters student
+
+
+#### Send speculative tasks to the workers
```

Comment:
I don't understand the change this proposal is making. Is it that when task `B` depends on task `A`, we assign task `B` to a machine before `A` is finished? And this is helpful because... (this is the part I don't get)

Reply:
I would like to remove the call to the low-level `fuse` optimization function.  However, in many workloads scheduling is better if I/O operations like `read_parquet` are immediately followed by the next sequence of tasks, like `.x.sum()` rather than waiting for those tasks to come from the scheduler.  (usually the worker starts on some other `read_parquet` task and fills up memory a bit faster otherwise).  

By having the subsequent tasks already on the worker we get the same advantages of low-level task fusion, but without actually having to fuse the tasks.

As an additional advantage, we'll separately account for costs of `read_parquet` and `add` and `sum` rather than have task prefixes like `read_parquet-add-sum`.